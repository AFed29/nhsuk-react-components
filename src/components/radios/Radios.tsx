import React, { HTMLProps, PureComponent, useState, useEffect, useContext } from 'react';
import classNames from 'classnames';
import { generateRandomName } from '../../util/RandomName';
import { LabelProps } from '../label/Label';
import { HintProps } from '../hint/Hint';
import LabelBlock from '../utils/LabelBlock';
import { ErrorMessageProps } from '../error-message/ErrorMessage';
import Label from '../label';
import RadioContext, { IRadioContext } from './RadioContext';
import Hint from '../hint';
import FormContext from '../form/FormContext';

interface RadioProps extends HTMLProps<HTMLInputElement> {
  labelProps?: HTMLProps<HTMLLabelElement>;
  hint?: string;
}

const Radio: React.FC<RadioProps> = ({
  className,
  children,
  id,
  labelProps,
  name,
  hint,
  ...rest
}) => {
  const [radioId, setRadioId] = useState<string | undefined>(undefined);
  const context = useContext<IRadioContext>(RadioContext);

  useEffect(() => {
    if (id) {
      setRadioId(id);
    } else if (context) {
      setRadioId(context.getRadioId());
    }
  }, [id]);

  return (
    <div className="nhsuk-radios__item" {...rest}>
      <input
        id={radioId}
        name={name || context.name}
        className={classNames('nhsuk-radios__input', className)}
        {...rest}
      />
      {children ? (
        <Label
          id={labelProps && labelProps.id ? labelProps.id : `${radioId}-label`}
          htmlFor={radioId}
          className={classNames(
            'nhsuk-radios__label',
            labelProps ? labelProps.className : undefined,
          )}
        >
          {children}
        </Label>
      ) : null}
      {hint ? (
        <Hint className="nhsuk-radios__hint" id={radioId ? `${radioId}-hint` : undefined}>
          {hint}
        </Hint>
      ) : null}
    </div>
  );
};

Radio.defaultProps = {
  type: 'radio',
};

const Divider: React.FC<HTMLProps<HTMLDivElement>> = ({ className, ...rest }) => (
  <div className={classNames('nhsuk-radios__divider', className)} {...rest} />
);

interface Radios {
  radioCount: number;
  name: string;
}

interface RadiosProps extends HTMLProps<HTMLDivElement> {
  inline?: boolean;
  label?: string;
  labelProps?: LabelProps;
  hint?: string;
  hintProps?: HintProps;
  error?: boolean | string;
  errorProps?: ErrorMessageProps;
}

class Radios extends PureComponent<RadiosProps> {
  static contextType = FormContext;

  static Radio = Radio;

  static Divider = Divider;

  constructor(props: RadiosProps, ...rest: any[]) {
    super(props, ...rest);
    this.name = props.name || generateRandomName('radios');
    this.radioCount = 0;
  }

  getRadioId = (): string | undefined => {
    const { id } = this.props;
    if (id) {
      this.radioCount += 1;
      return `${id}-${this.radioCount}`;
    }
    return undefined;
  };

  render() {
    const {
      className,
      id,
      children,
      inline,
      label,
      labelProps,
      hint,
      hintProps,
      error,
      errorProps,
      ...rest
    } = this.props;

    const { isForm, setError } = this.context;
    if (isForm) {
      setError(this.name, Boolean(error));
    }

    return (
      <>
        <LabelBlock
          elementId={id}
          label={label}
          labelProps={labelProps}
          hint={hint}
          hintProps={hintProps}
          error={error}
          errorProps={errorProps}
        />
        <RadioContext.Provider value={{ getRadioId: this.getRadioId, name: this.name }}>
          <div
            className={classNames('nhsuk-radios', { 'nhsuk-radios--inline': inline }, className)}
            id={id}
            {...rest}
          >
            {children}
          </div>
        </RadioContext.Provider>
      </>
    );
  }
}

export default Radios;
