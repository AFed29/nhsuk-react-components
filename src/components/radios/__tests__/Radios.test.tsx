import React from 'react';
import { mount } from 'enzyme';
import Radios from '../Radios';

describe('Radios', () => {
  it('matches snapshot', () => {
    const element = mount(
      <Radios id="example" name="example">
        <Radios.Radio id="example-1" value="yes">
          Yes
        </Radios.Radio>
        <Radios.Radio id="example-2" value="no">
          No
        </Radios.Radio>
      </Radios>,
    );
    expect(element).toMatchSnapshot();
    element.unmount();
  });

  it('does not render the conditional content if checked is false', () => {
    const element = mount(
      <Radios id="example" name="example">
        <Radios.Radio id="example-1" value="yes" checked={false} conditional={<p>Test</p>}>
          Yes
        </Radios.Radio>
        <Radios.Radio id="example-2" value="no">
          No
        </Radios.Radio>
      </Radios>,
    );
    element.find('input#example-1').simulate('change');
    expect(element).toMatchSnapshot();
    element.unmount();
  });

  it('renders the conditional content if the radio reference = selected radio', () => {
    const element = mount(
      <Radios id="example" name="example">
        <Radios.Radio id="example-1" value="yes" conditional={<p>Test</p>}>
          Yes
        </Radios.Radio>
        <Radios.Radio id="example-2" value="no">
          No
        </Radios.Radio>
      </Radios>,
    );
    element.find('input#example-1').simulate('change');
    expect(element).toMatchSnapshot();
    element.unmount();
  });
});
