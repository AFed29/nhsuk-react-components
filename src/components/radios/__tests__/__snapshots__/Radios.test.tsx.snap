// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radios matches snapshot 1`] = `
<Radios
  id="example"
  name="example"
  role="radiogroup"
>
  <FormGroup
    id="example"
    inputType="radios"
    name="example"
    role="radiogroup"
  >
    <div
      className="nhsuk-form-group"
    >
      <div
        className="nhsuk-radios"
        id="example"
        role="radiogroup"
      >
        <Radio
          id="example-1"
          type="radio"
          value="yes"
        >
          <div
            className="nhsuk-radios__item"
          >
            <input
              aria-labelledby="example-1--label"
              className="nhsuk-radios__input"
              id="example-1"
              name="example"
              onChange={[Function]}
              type="radio"
              value="yes"
            />
            <Label
              className="nhsuk-radios__label"
              htmlFor="example-1"
              id="example-1--label"
            >
              <BaseLabel
                className="nhsuk-radios__label"
                htmlFor="example-1"
                id="example-1--label"
              >
                <label
                  className="nhsuk-label nhsuk-radios__label"
                  htmlFor="example-1"
                  id="example-1--label"
                >
                  Yes
                </label>
              </BaseLabel>
            </Label>
          </div>
        </Radio>
        <Radio
          id="example-2"
          type="radio"
          value="no"
        >
          <div
            className="nhsuk-radios__item"
          >
            <input
              aria-labelledby="example-2--label"
              className="nhsuk-radios__input"
              id="example-2"
              name="example"
              onChange={[Function]}
              type="radio"
              value="no"
            />
            <Label
              className="nhsuk-radios__label"
              htmlFor="example-2"
              id="example-2--label"
            >
              <BaseLabel
                className="nhsuk-radios__label"
                htmlFor="example-2"
                id="example-2--label"
              >
                <label
                  className="nhsuk-label nhsuk-radios__label"
                  htmlFor="example-2"
                  id="example-2--label"
                >
                  No
                </label>
              </BaseLabel>
            </Label>
          </div>
        </Radio>
      </div>
    </div>
  </FormGroup>
</Radios>
`;

exports[`Radios does not render the conditional content if checked is false 1`] = `
<Radios
  id="example"
  name="example"
  role="radiogroup"
>
  <FormGroup
    id="example"
    inputType="radios"
    name="example"
    role="radiogroup"
  >
    <div
      className="nhsuk-form-group"
    >
      <div
        className="nhsuk-radios nhsuk-radios--conditional"
        id="example"
        role="radiogroup"
      >
        <Radio
          checked={false}
          conditional={
            <p>
              Test
            </p>
          }
          id="example-1"
          type="radio"
          value="yes"
        >
          <div
            className="nhsuk-radios__item"
          >
            <input
              aria-labelledby="example-1--label"
              checked={false}
              className="nhsuk-radios__input"
              id="example-1"
              name="example"
              onChange={[Function]}
              type="radio"
              value="yes"
            />
            <Label
              className="nhsuk-radios__label"
              htmlFor="example-1"
              id="example-1--label"
            >
              <BaseLabel
                className="nhsuk-radios__label"
                htmlFor="example-1"
                id="example-1--label"
              >
                <label
                  className="nhsuk-label nhsuk-radios__label"
                  htmlFor="example-1"
                  id="example-1--label"
                >
                  Yes
                </label>
              </BaseLabel>
            </Label>
          </div>
        </Radio>
        <Radio
          id="example-2"
          type="radio"
          value="no"
        >
          <div
            className="nhsuk-radios__item"
          >
            <input
              aria-labelledby="example-2--label"
              className="nhsuk-radios__input"
              id="example-2"
              name="example"
              onChange={[Function]}
              type="radio"
              value="no"
            />
            <Label
              className="nhsuk-radios__label"
              htmlFor="example-2"
              id="example-2--label"
            >
              <BaseLabel
                className="nhsuk-radios__label"
                htmlFor="example-2"
                id="example-2--label"
              >
                <label
                  className="nhsuk-label nhsuk-radios__label"
                  htmlFor="example-2"
                  id="example-2--label"
                >
                  No
                </label>
              </BaseLabel>
            </Label>
          </div>
        </Radio>
      </div>
    </div>
  </FormGroup>
</Radios>
`;

exports[`Radios renders the conditional content if the radio reference = selected radio 1`] = `
<Radios
  id="example"
  name="example"
  role="radiogroup"
>
  <FormGroup
    id="example"
    inputType="radios"
    name="example"
    role="radiogroup"
  >
    <div
      className="nhsuk-form-group"
    >
      <div
        className="nhsuk-radios nhsuk-radios--conditional"
        id="example"
        role="radiogroup"
      >
        <Radio
          conditional={
            <p>
              Test
            </p>
          }
          id="example-1"
          type="radio"
          value="yes"
        >
          <div
            className="nhsuk-radios__item"
          >
            <input
              aria-labelledby="example-1--label"
              className="nhsuk-radios__input"
              id="example-1"
              name="example"
              onChange={[Function]}
              type="radio"
              value="yes"
            />
            <Label
              className="nhsuk-radios__label"
              htmlFor="example-1"
              id="example-1--label"
            >
              <BaseLabel
                className="nhsuk-radios__label"
                htmlFor="example-1"
                id="example-1--label"
              >
                <label
                  className="nhsuk-label nhsuk-radios__label"
                  htmlFor="example-1"
                  id="example-1--label"
                >
                  Yes
                </label>
              </BaseLabel>
            </Label>
          </div>
          <div
            className="nhsuk-radios__conditional"
            id="example-1--conditional"
          >
            <p>
              Test
            </p>
          </div>
        </Radio>
        <Radio
          id="example-2"
          type="radio"
          value="no"
        >
          <div
            className="nhsuk-radios__item"
          >
            <input
              aria-labelledby="example-2--label"
              className="nhsuk-radios__input"
              id="example-2"
              name="example"
              onChange={[Function]}
              type="radio"
              value="no"
            />
            <Label
              className="nhsuk-radios__label"
              htmlFor="example-2"
              id="example-2--label"
            >
              <BaseLabel
                className="nhsuk-radios__label"
                htmlFor="example-2"
                id="example-2--label"
              >
                <label
                  className="nhsuk-label nhsuk-radios__label"
                  htmlFor="example-2"
                  id="example-2--label"
                >
                  No
                </label>
              </BaseLabel>
            </Label>
          </div>
        </Radio>
      </div>
    </div>
  </FormGroup>
</Radios>
`;
